src/envoy/http/authn/http_filter.cc

```diff
@@ -14,75 +14,65 @@
+
+  authenticator_ = std::move(createPeerAuthenticator(
+      filter_context_.get(),
+      [this](bool success) { onPeerAuthenticationDone(success); }));
+  authenticator_->run();
 
-  int peer_size = config_.peers_size();
-  ENVOY_LOG(debug, "AuthenticationFilter: {} config.peers_size()={}", __func__,
-            peer_size);
-  if (peer_size > 0) {
-    const auto& m = config_.peers()[0];
-    if (m.has_mtls()) {
-      ENVOY_LOG(debug, "AuthenticationFilter: {} this connection requires mTLS",
-                __func__);
-      MtlsAuthentication mtls_authn(decoder_callbacks_->connection());
-      if (mtls_authn.IsMutualTLS() == false) {
-        // In prototype, only log the authentication policy violation.
-        ENVOY_LOG(error,
-                  "AuthenticationFilter: authn policy requires mTLS but the "
-                  "connection is not mTLS!");
-      } else {
-        ENVOY_LOG(debug, "AuthenticationFilter: the connection is mTLS.");
-        std::string user, ip;
-        int port = 0;
-        bool ret = false;
-        ret = mtls_authn.GetSourceUser(&user);
-        if (ret) {
-          ENVOY_LOG(debug, "AuthenticationFilter: the source user is {}", user);
-        } else {
-          ENVOY_LOG(error,
-                    "AuthenticationFilter: GetSourceUser() returns false!");
-        }
-        ret = mtls_authn.GetSourceIpPort(&ip, &port);
-        if (ret) {
-          ENVOY_LOG(debug,
-                    "AuthenticationFilter: the source ip is {}, the source "
-                    "port is {}",
-                    user, port);
-        } else {
-          ENVOY_LOG(error,
-                    "AuthenticationFilter: GetSourceIpPort() returns false!");
-        }
-      }
-    } else {
-      ENVOY_LOG(
-          debug,
-          "AuthenticationFilter: {} this connection does not require mTLS",
-          __func__);
-    }
+  if (state_ == Istio::AuthN::State::COMPLETE) {
+    return FilterHeadersStatus::Continue;
   }
 
-  ENVOY_LOG(
-      debug,
-      "Called AuthenticationFilter : {}, return FilterHeadersStatus::Continue;",
-      __func__);
-  return FilterHeadersStatus::Continue;
+  stopped_ = true;
+  return FilterHeadersStatus::StopIteration;
+}
+
+void AuthenticationFilter::onPeerAuthenticationDone(bool success) {
+  ENVOY_LOG(debug, "{}: success = {}", __func__, success);
+  if (success) {
+    authenticator_ = std::move(createOriginAuthenticator(
+        filter_context_.get(),
+        [this](bool success) { onOriginAuthenticationDone(success); }));
+    authenticator_->run();
+  } else {
+    rejectRequest("Peer authentication failed.");
+  }
+}
+void AuthenticationFilter::onOriginAuthenticationDone(bool success) {
+  ENVOY_LOG(debug, "{}: success = {}", __func__, success);
+  if (success) {
+    continueDecoding();
+  } else {
+    rejectRequest("Origin authentication failed.");
+  }
 }
 
 FilterDataStatus AuthenticationFilter::decodeData(Buffer::Instance&, bool) {
```

Comment:
at decodeData() 

you need to check state_ too.

Reply:
Done